source/sigv4.c

```diff
@@ -1814,10 +1819,69 @@ static void generateCredentialScope( const SigV4Parameters_t * pSigV4Params,
 
 /*-----------------------------------------------------------*/
 
-    static SigV4Status_t setQueryStringFieldsAndValues( const char * pQuery,
-                                                        size_t queryLen,
-                                                        size_t * pNumberOfParameters,
-                                                        CanonicalContext_t * pCanonicalRequest )
+    static void processAmpersantInQueryString( bool fieldHasValue,
+                                               size_t currQueryIndex,
+                                               size_t startOfFieldOrValue,
+                                               size_t * pCurrParamCount,
+                                               const char * pQuery,
+                                               CanonicalContext_t * pCanonicalRequest )
+    {
+        bool previousParamIsValid = true;
+
+        assert( pCurrParamCount != NULL );
+        assert( pQuery != NULL );
+        assert( pCanonicalRequest != NULL );
+
+        /* Ignore unnecessary '&' that represent empty parameter names.
+         * Trimmable '&'s can be leading, trailing or repeated as separators between parameter
+         * pairs.
+         * For example, this function will prune "&p1=v1&&p2=v2&" to "p1=v1&p2=v2". */
+        if( !( fieldHasValue ) && ( ( ( currQueryIndex - startOfFieldOrValue ) == 0U ) ) )
+        {
+            /* Do nothing as previous parameter name is empty. */
+            previousParamIsValid = false;
+        }
+
+        /* Case when parameter has empty value even though the query parameter entry has the form
+         * "<param>=". */
+        else if( ( fieldHasValue ) && ( ( currQueryIndex - startOfFieldOrValue ) == 0U ) )
```

Comment:
So this will happen in the case of either `<param>=<end of query>` or `<param>=&`. Are either of these valid?

Reply:
The [API Gateway Tutorial](https://docs.aws.amazon.com/apigateway/latest/developerguide/api-gateway-create-api-step-by-step.html#getting-started-mappings) provides an example that supports providing query parameters with no value in both the formats. 
The example REST API is: http://petstore-demo-endpoint.execute-api.com/petstore/pets?type=cat&page=2

Reply:
Moreover, the [AWS SigV4 spec](https://docs.aws.amazon.com/general/latest/gr/sigv4-create-canonical-request.html) mentions about providing the query parameter name with `=` character but an empty string for empty values.

> For each parameter, append the URI-encoded parameter name, followed by the equals sign character (=), followed by the URI-encoded parameter value. Use an empty string for parameters that have no value.